File: Dynamic Programming/Hard/LET_no85_Maximal Rectangle.py
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        m = len(matrix)
        if not m:
            return 0
        n = len(matrix[0])
        left, right, height = [0] * n, [n] * n, [0] * n
        res = 0
        
        for i in range(m):
            curLeft, curRight = 0, n
            for j in range(n):
                if matrix[i][j] == '1':
                    left[j] = max(left[j], curLeft)
                    height[j] += 1
                else:
                    left[j] = 0
                    curLeft = j + 1
                    height[j] = 0
            for j in range(n - 1, -1, -1):
                if matrix[i][j] == '1':
                    right[j] = min(right[j], curRight)
                    res = max(res, (right[j] - left[j]) * height[j])
                else:
                    right[j] = n
                    curRight = j
        
        return res

File: Dynamic Programming/Hard/LET_no85_Maximal Rectangle.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int m = matrix.size();
        if (m == 0)
            return 0;
        int n = matrix[0].size();
        vector<int> left(n, 0), right(n, n), height(n, 0);
        int res = 0;
        
        for (int i = 0; i < m; i++) {
            int curLeft = 0, curRight = n;
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == '1') {
                    left[j] = max(left[j], curLeft);
                    height[j]++;
                }
                else {
                    height[j] = left[j] = 0;
                    curLeft = j + 1;
                }
            }
            for (int j = n - 1; j >= 0; j--) {
                if (matrix[i][j] == '1') {
                    right[j] = min(right[j], curRight);
                    res = max(res, (right[j] - left[j]) * height[j]);
                }
                else {
                    right[j] = n;
                    curRight = j;
                }
            }
        }
        
        return res;
    }
};